plutus-core/plutus-core/src/Universe/Core.hs

```diff
@@ -0,0 +1,756 @@
+in client code but without the type checker forcing us to do so like it does now, so that's a huge
+price to pay for some superficial syntactic nicety and hence we choose the safe approach,
+even though that required reworking all the infrastructure in a backwards-incompatible manner.
+-}
+
+-- See Note [Representing polymorphism].
+type T :: forall k. k -> Type
+data T a
+
+-- | Existential quantification as a data type.
+type Some :: forall a. (a -> Type) -> Type
+data Some f = forall x. Some !(f x)
+
+-- | A particular type from a universe.
+type SomeTypeIn :: (Type -> Type) -> Type
+data SomeTypeIn uni = forall k (a :: k). SomeTypeIn !(uni (T a))
+
+data Kinded uni ta where
+    Kinded :: Typeable k => !(uni (T a)) -> Kinded uni (T (a :: k))
+
+-- | A value of a particular type from a universe.
+type ValueOf :: (Type -> Type) -> Type -> Type
+data ValueOf uni a = ValueOf !(uni (T a)) !a
+
+-- | A class for enumerating types and fully instantiated type formers that @uni@ contains.
+-- For example, a particular @ExampleUni@ may have monomorphic types in it:
+--
+--     instance ExampleUni `Contains` Integer where <...>
+--     instance ExampleUni `Contains` Bool    where <...>
+--
+-- as well as polymorphic ones:
+--
+--     instance ExampleUni `Contains` [] where <...>
+--     instance ExampleUni `Contains` (,) where <...>
+--
+-- as well as their instantiations:
+--
+--     instance ExampleUni `Contains` a => ExampleUni `Contains` [a] where <...>
+--     instance (ExampleUni `Contains` a, ExampleUni `Contains` b) => ExampleUni `Contains` (a, b) where <...>
+--
+-- (a universe can have any subset of the mentioned sorts of types, for example it's fine to have
+-- instantiated polymorphic types and not have uninstantiated ones and vice versa)
+--
+-- Note that when used as a constraint of a function 'Contains' does not allow you to directly
+-- express things like \"@uni@ has the @Integer@, @Bool@ and @[]@ types and type formers\",
+-- because @[]@ is not fully instantiated. So you can only say \"@uni@ has @Integer@, @Bool@,
+-- @[Integer]@, @[Bool]@, @[[Integer]]@, @[[Bool]]@ etc\" and such manual enumeration is annoying,
+-- so we'd really like to be able to say that @uni@ has lists of arbitrary built-in types
+-- (including lists of lists etc). 'Contains' does not allow that, but 'Includes' does.
+-- For example, in the body of the following definition:
+--
+--     foo :: (uni `Includes` Integer, uni `Includes` Bool, uni `Includes` []) => <...>
+--     foo = <...>
+--
+-- you can make use of the fact that @uni@ has lists of arbitrary included types (integers,
+-- booleans and lists).
+--
+-- Hence most of the time opt for using the more flexible 'Includes'.
+--
+-- 'Includes' is defined in terms of 'Contains', so you only need to provide a 'Contains' instance
+-- per type from the universe and you'll get 'Includes' for free.
+type Contains :: forall k. (Type -> Type) -> k -> Constraint
+class uni `Contains` a where
+    knownUni :: uni (T a)
```

Comment:
This is an odd name, maybe? If we had `type Tag uni a = uni (T a)` then maybe that suggests this should be `knownTag`? or just `tag`? Or some other word for tag (`UniTag`?)

Reply:
So I use `uni` everywhere to mean either "a universe" or "a type from the universe" depending on whether we're at the type or the term level. Analogously to `fun` meaning either "a set of built-in functions" or "a built-in function from the set of built-in functions" depending on whether we're at the type or the term level. I could make it `tag`, but I like that `uni` is more specific, even if it can mean two different things depending on where we are (which I think is pretty much never a problem just like calling a `newtype` an its constructor the same way is pretty much never a problem). Anyway, we could do that, but then it's going to touch literally everything related to universes and I'd prefer to do that in a separate PR (or something else can do it once this PR is merged).

Reply:
I.e. I treat "uni" at the term level as a shorthand for "type tag".

Reply:
Okay, well _I_ find that punning confusing. I think I've been confused about this in the past too. I also find the distinction between types in a universe and tags in a universe sometimes confusing. Maybe this is no problem when you're used to it, I don't know if it's worth optimizing legibility for people who would be confused by this.

In the case of the tags, my confusion basically boils down to seeing `x :: uni (T a)` and going "x is a universe? what?" and then remembering that it doesn't work like that. Basically I learned to read `uni (T a)` as "tag". But then why not just help me and write `Tag`?

Reply:
> I also find the distinction between types in a universe and tags in a universe sometimes confusing.

Semantics versus syntax pretty much. As the docs say:

> A universe is a collection of tags for types. It can be finite like
>
>             data U a where   
>                 UUnit :: U ()      
>                 UInt  :: U Int
> 
> (where 'UUnit' is a tag for '()' and 'UInt' is a tag for 'Int') or <...>
>
> We say that a type is in a universe whenever there is a tag for that type in the universe.

so I think this particular distinction is well-specified. But maybe now I need to clarify that a tag is not necessary a single constructor and can be an application of a constructor to another constructor/application.

> In the case of the tags, my confusion basically boils down to seeing x :: uni (T a) and going "x is a universe? what?"

We'd had `x :: uni a` up until this PR and it always meant "a type tag for `a` in `uni`". `x` never was a universe.

> Basically I learned to read `uni (T a)` as "tag". But then why not just help me and write `Tag`?

Right, that does seem better, I'll try it out.

Reply:
> We'd had x :: uni a up until this PR and it always meant "a type tag for a in uni". x never was a universe.

Yeah, and I think I was at least somewhat confused every time I saw it :D